caffe2/core/operator_c10wrapper.h

```diff
@@ -25,104 +16,100 @@ using extract_type_t = typename ParameterDef::type;
   }
 
   bool RunOnDevice() override {
     // due to caching the stack_, concurrent calling is not allowed.
     // TODO thread_local might fix this
     std::lock_guard<std::mutex> lock(mutex_);
 
-    AT_ASSERT(stack_.size() == 0);
-
     pushInputs_();
-    pushParameters_(guts::make_index_sequence<num_parameters()>());
-    pushOutputParameters_();
-
     callKernel_();
-
     popOutputs_();
 
     return true;
   }
 
  private:
-  template <size_t... ParameterIndex>
-  ParameterTuple parse_parameters_(
-      const OperatorDef& operator_def,
-      c10::guts::index_sequence<ParameterIndex...>) {
-    return ParameterTuple{Parameter<ParameterIndex>(operator_def)...};
-  }
-
-  template <size_t Index>
-  details::extract_type_t<
-      typename std::tuple_element<Index, ParameterDefTuple>::type>
-  Parameter(const OperatorDef& operator_def) {
-    using Parameter =
-        typename std::tuple_element<Index, ParameterDefTuple>::type;
-    return Parameter::parse(ArgumentHelper(operator_def));
-  }
-
   void pushInputs_() {
-    if (use_array_input) {
-      stack_.emplace_back(ivalue::TensorList::create(array_inputs_()));
-    } else {
-      for (size_t i = 0; i < num_inputs(); ++i) {
-        stack_.emplace_back(at::Tensor(C10Tensor(Input(i))));
+    AT_ASSERT(stack_.size() == 0);
+    stack_.reserve(
+        op_.schema().arguments().size() + (has_preallocated_outputs_ ? 1 : 0));
+
+    size_t input_tensor_index = 0;
+
+    for (const auto& argument : op_.schema().arguments()) {
+      if (argument.name() == detail::PREALLOCATED_OUTPUT_ARGNAME) {
+        // note: if detail::PREALLOCATED_OUTPUT_ARGNAME was at the end of the
+        // argument list, then has_preallocated_outputs_ would be true.
+        AT_ASSERTM(
+            has_preallocated_outputs_,
+            "Error in caffe2->c10 wrapper: Operator schema has a parameter named ",
+            detail::PREALLOCATED_OUTPUT_ARGNAME,
+            ", but it's not at the end of the argument list");
+
+        AT_ASSERTM(
+            argument.type()->isSubtypeOf(
+                OptionalType::create(ListType::ofTensors())),
+            "Error in caffe2->c10 wrapper: Operator schema has a parameter named ",
+            detail::PREALLOCATED_OUTPUT_ARGNAME,
+            ", but it's not of type TensorList?");
+        stack_.emplace_back(preallocated_outputs_());
+
+      } else if (argument.type()->isSubtypeOf(TensorType::get())) {
+        AT_ASSERTM(
+            input_tensor_index < InputSize(),
+            "Error in caffe2->c10 wrapper: Too few tensor arguments given (",
+            InputSize(),
+            "), operator schema expected more.");
+        stack_.emplace_back(at::Tensor(Input(input_tensor_index++)));
+
+      } else if (argument.type()->isSubtypeOf(ListType::ofTensors())) {
+        AT_ASSERTM(
+            input_tensor_index == 0,
+            "Error in caffe2->c10 wrapper: Schema can only have either one or more Tensor inputs or one TensorList input.");
```

Comment:
You're not actually checking this invariant, though, if the argument order is `TensorList, Tensor`, is that right?

Reply:
This case sets `input_tensor_index = InputSize()` at line 100, which then fails the assertion in line 88 if an argument of type `Tensor` follows.